radix-engine/src/model/resource.rs

```diff
@@ -0,0 +1,282 @@
+use sbor::*;
+use scrypto::engine::types::*;
+use scrypto::rust::collections::BTreeMap;
+use scrypto::rust::collections::BTreeSet;
+use scrypto::rust::collections::HashMap;
+use scrypto::rust::rc::Rc;
+use scrypto::rust::string::ToString;
+
+/// Represents some error when manipulating resources.
+#[derive(Debug, Clone)]
+pub enum ResourceError {
+    /// Resource addresses do not match
+    ResourceAddressNotMatching,
+    /// The amount is invalid, according to the resource divisibility
+    InvalidAmount(Decimal, u8),
+    /// The balance is not enough
+    InsufficientBalance,
+    /// The operation is not supported
+    UnsupportedOperation,
+}
+
+/// Represents the contained resource.
+#[derive(Debug, Clone, TypeId, Encode, Decode)]
+pub enum Resource {
+    // TODO: update resource state based on proofs.
+    Fungible {
+        /// The locked amounts and the corresponding times of being locked.
+        locked_amounts: BTreeMap<Decimal, usize>,
+        /// The liquid amount.
+        liquid_amount: Decimal,
+    },
+    NonFungible {
+        /// The locked non-fungible ids and the corresponding times of being locked.
+        locked_ids: HashMap<NonFungibleId, usize>,
+        /// The liquid non-fungible ids.
+        liquid_ids: BTreeSet<NonFungibleId>,
+    },
+}
+
+#[derive(Debug, Clone, TypeId, Encode, Decode)]
+pub enum ResourceAmount {
+    /// Fungible amount
+    Fungible { amount: Decimal },
+    /// Non-fungible amount
+    NonFungible { ids: BTreeSet<NonFungibleId> },
+}
+
+/// A transient resource container.
+#[derive(Debug, TypeId, Encode, Decode)]
+pub struct ResourceContainer {
+    /// The resource definition id
+    resource_def_id: ResourceDefId,
+    /// The resource type
+    resource_type: ResourceType,
+    /// The contained resource
+    resource: Resource,
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum ResourceContainerId {
+    /// For named bucket
+    Bucket(BucketId),
+    /// For vault
+    Vault(VaultId),
+
+    /// For the specific resource on the n-th worktop
+    Worktop {
+        depth: u32,
+        resource_def_id: ResourceDefId,
+    },
+}
+
+#[derive(Debug, Clone)]
+pub struct Proof {
+    /// The resource definition id
+    resource_def_id: ResourceDefId,
+    /// The resource type
+    resource_type: ResourceType,
+    /// Restricted proof can't be moved down along the call stack (growing down).
+    restricted: bool,
+    /// The total amount locked
+    total_amount: ResourceAmount,
```

Comment:
Is the `total_amount` here the sum of all the values in `amounts`? If so, should add a comment that this is optimization to get total_amount faster.

Reply:
Yes, this is purely for optimization. It can be derived from the hashmap below. I will add comments.